ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/explorer/project/synchronize/SelectionAnalyzer.java

```diff
@@ -0,0 +1,138 @@
+import org.eclipse.che.api.promises.client.Promise;
+import org.eclipse.che.api.workspace.shared.dto.SourceStorageDto;
+import org.eclipse.che.ide.CoreLocalizationConstant;
+import org.eclipse.che.ide.api.app.AppContext;
+import org.eclipse.che.ide.api.dialogs.DialogFactory;
+import org.eclipse.che.ide.api.event.SelectionChangedEvent;
+import org.eclipse.che.ide.api.event.SelectionChangedHandler;
+import org.eclipse.che.ide.api.notification.NotificationManager;
+import org.eclipse.che.ide.api.resources.Project;
+import org.eclipse.che.ide.api.resources.marker.Marker;
+import org.eclipse.che.ide.util.loging.Log;
+
+import java.util.Map;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static org.eclipse.che.ide.api.notification.StatusNotification.DisplayMode.EMERGE_MODE;
+import static org.eclipse.che.ide.api.notification.StatusNotification.Status.SUCCESS;
+import static org.eclipse.che.ide.api.resources.Project.ProblemProjectMarker.PROBLEM_PROJECT;
+
+/**
+ * Synchronize projects in workspace with projects on file system.
+ * The synchronization performs based on project problems.
+ * Project problem with code 10 means that project exist in workspace but it is absent on file system.
+ */
+@Singleton
+public class SelectionAnalyzer implements SelectionChangedHandler {
+    private final AppContext               appContext;
+    private final DialogFactory            dialogFactory;
+    private final CoreLocalizationConstant locale;
+    private final NotificationManager      notificationManager;
+    private final ChangeLocationWidget     changeLocationWidget;
+
+    @Inject
+    public SelectionAnalyzer(EventBus eventBus,
+                             AppContext appContext,
+                             DialogFactory dialogFactory,
+                             CoreLocalizationConstant locale,
+                             NotificationManager notificationManager,
+                             ChangeLocationWidget changeLocationWidget) {
+        this.appContext = appContext;
+        this.dialogFactory = dialogFactory;
+        this.locale = locale;
+        this.notificationManager = notificationManager;
+        this.changeLocationWidget = changeLocationWidget;
+
+        eventBus.addHandler(SelectionChangedEvent.TYPE, this);
+    }
+
+    @Override
+    public void onSelectionChanged(SelectionChangedEvent event) {
+        final Project project = appContext.getRootProject();
+        if (project == null) {
+            return;
+        }
+
+        final Optional<Marker> marker = project.getMarker(PROBLEM_PROJECT);
+        if (!marker.isPresent()) {
+            return;
+        }
+
+        final Project.ProblemProjectMarker problemProjectMarker = (Project.ProblemProjectMarker)marker.get();
+        final Map<Integer, String> problems = problemProjectMarker.getProblems();
+
+        //If no project folder on file system
+        final String noProjectFolderProblem = problems.get(10);
```

Comment:
I think we need somewhere represent an object or enum with possible error codes, which server provides. Not related to this issue, but anyway